### PR TITLE
create profiles contract and some tests

### PR DESCRIPTION
**Pull request by ramirocarra**

### CODE REVIEW

File: contracts/Profiles.sol [hunk 1]
```diff
@@ -0,0 +1,348 @@
...
+  uint256[] public experienceTable = [0, 10, 15, 20, 30, 40, 60];
+
+  mapping(uint256 => Profile) public profiles;
+  mapping(string => uint256) private _namesToProfileId;
+  mapping(address => bool) private _avatarContractsEnabled;
+
+  uint8 private _maxDefaultAvatars = 4;
+  string private _defaultAvatarBaseURI = "https://terrae.finance/avatars/";
+
+  uint32 public maxStamina = 100;
```
Comment (scucchiero) [inline contracts/Profiles.sol:29]: stamina -> energy

File: contracts/Profiles.sol [hunk 2]
```diff
@@ -0,0 +1,348 @@
...
+  }
+
+  // Avoid transfer to users that already own a profile
+  function _beforeTokenTransfer(
+    address from,
+    address to,
+    uint256 amount
+  ) internal virtual override {
+    super._beforeTokenTransfer(from, to, amount);
+    require(balanceOf(to) == 0, "Recieving address already owns a profile");
```
Comment (scucchiero) [inline contracts/Profiles.sol:65]: typo

File: contracts/Profiles.sol [hunk 3]
```diff
@@ -0,0 +1,348 @@
...
+  constructor() ERC721("TerraeProfile", "TPC") {
+    _setupRole(DEFAULT_ADMIN_ROLE, msg.sender);
+    _setupRole(MANTAINER_ROLE, msg.sender);
+
+    // assign profile zero to owner
+    createProfile("Owner", 0);
+  }
+
+  function _baseURI() internal pure override returns (string memory) {
+    return "https://terrae.finance/profile/";
```
Comment (scucchiero) [inline contracts/Profiles.sol:55]: TBD

File: contracts/Profiles.sol (hunk 2 above)
Comment (scucchiero) [reply contracts/Profiles.sol:65]: also we can use require(!hasProfile(to)

File: contracts/Profiles.sol [hunk 4]
```diff
@@ -0,0 +1,348 @@
...
+  bytes32 private constant USE_STAMINA_ROLE = keccak256("USE_STAMINA_ROLE");
+  bytes32 private constant MANTAINER_ROLE = keccak256("MANTAINER_ROLE");
+  bytes32 private constant ADD_EXP_ROLE = keccak256("ADD_EXP_ROLE");
+  // Burning profiles??
+
+  //upgradable? how much will this cost?
+  uint256[] public experienceTable = [0, 10, 15, 20, 30, 40, 60];
+
+  mapping(uint256 => Profile) public profiles;
+  mapping(string => uint256) private _namesToProfileId;
```
Comment (scucchiero) [inline contracts/Profiles.sol:23]: correct to _nameToProfileId

File: contracts/Profiles.sol [hunk 5]
```diff
@@ -0,0 +1,348 @@
...
+  uint8 private _maxDefaultAvatars = 4;
+  string private _defaultAvatarBaseURI = "https://terrae.finance/avatars/";
+
+  uint32 public maxStamina = 100;
+  uint32 public secondsPerStamina = 300;
+
+  struct Profile {
+    string name; //limits???? moderator?
+    uint256 created;
+    uint32 exp;
```
Comment (scucchiero) [inline contracts/Profiles.sol:35]: use only uint256 for all uints

File: contracts/Profiles.sol [hunk 6]
```diff
@@ -0,0 +1,348 @@
...
+   * @dev Checks if the address has a profile.
+   */
+  function hasProfile(address _address) public view returns (bool) {
+    return balanceOf(_address) > 0;
+  }
+
+  /**
+   * @dev Checks if the name is already taken.
+   */
+  function nameTaken(string memory _name) public view returns (bool taken) {
```
Comment (scucchiero) [inline contracts/Profiles.sol:88]: we dont need to set a name for the variable in return value 
change to `returns (bool)`

File: contracts/Profiles.sol [hunk 7]
```diff
@@ -0,0 +1,348 @@
...
+import "@openzeppelin/contracts/utils/Arrays.sol";
+
+contract Profiles is AccessControl, ERC721Enumerable {
+  bytes32 private constant USE_STAMINA_ROLE = keccak256("USE_STAMINA_ROLE");
+  bytes32 private constant MANTAINER_ROLE = keccak256("MANTAINER_ROLE");
+  bytes32 private constant ADD_EXP_ROLE = keccak256("ADD_EXP_ROLE");
+  // Burning profiles??
+
+  //upgradable? how much will this cost?
+  uint256[] public experienceTable = [0, 10, 15, 20, 30, 40, 60];
```
Comment (scucchiero) [inline contracts/Profiles.sol:20]: we dont need this. 
we can work with exp points.
the concept of levels can be defined OFF chain

**Review by scucchiero — CHANGES_REQUESTED** [review]

File: contracts/Profiles.sol (hunk 1 above)
Comment (fedecaccia) [reply contracts/Profiles.sol]: can we call it elixir?

**Pull request closed without merging**